lightning/src/offers/offer.rs

```diff
@@ -0,0 +1,695 @@
+	///
+	/// See [`Offer::chains`] on how this relates to the payment currency.
+	///
+	/// Successive calls to this method will add another chain hash.
+	pub fn chain(mut self, network: Network) -> Self {
+		let chains = self.offer.chains.get_or_insert_with(Vec::new);
+		let chain = ChainHash::using_genesis_block(network);
+		if !chains.contains(&chain) {
+			chains.push(chain);
+		}
+
+		self
+	}
+
+	/// Sets the [`Offer::metadata`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn metadata(mut self, metadata: Vec<u8>) -> Self {
+		self.offer.metadata = Some(metadata);
+		self
+	}
+
+	/// Sets the [`Offer::amount`] as an [`Amount::Bitcoin`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn amount_msats(mut self, amount_msats: u64) -> Self {
+		self.amount(Amount::Bitcoin { amount_msats })
+	}
+
+	/// Sets the [`Offer::amount`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	fn amount(mut self, amount: Amount) -> Self {
+		self.offer.amount = Some(amount);
+		self
+	}
+
+	/// Sets the [`Offer::features`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	#[cfg(test)]
+	pub fn features(mut self, features: OfferFeatures) -> Self {
+		self.offer.features = features;
+		self
+	}
+
+	/// Sets the [`Offer::absolute_expiry`] as seconds since the Unix epoch. Any expiry that has
+	/// already passed is valid and can be checked for using [`Offer::is_expired`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn absolute_expiry(mut self, absolute_expiry: Duration) -> Self {
+		self.offer.absolute_expiry = Some(absolute_expiry);
+		self
+	}
+
+	/// Sets the [`Offer::issuer`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn issuer(mut self, issuer: String) -> Self {
+		self.offer.issuer = Some(issuer);
+		self
+	}
+
+	/// Adds a blinded path to [`Offer::paths`]. Must include at least one path if only connected by
+	/// private channels or if [`Offer::signing_pubkey`] is not a public node id.
+	///
+	/// Successive calls to this method will add another blinded path. Caller is responsible for not
+	/// adding duplicate paths.
+	pub fn path(mut self, path: BlindedPath) -> Self {
+		self.offer.paths.get_or_insert_with(Vec::new).push(path);
+		self
+	}
+
+	/// Sets the quantity of items for [`Offer::supported_quantity`].
+	///
+	/// Successive calls to this method will override the previous setting.
+	pub fn supported_quantity(mut self, quantity: Quantity) -> Self {
+		self.offer.supported_quantity = quantity;
+		self
+	}
+
+	/// Builds an [`Offer`] from the builder's settings.
+	pub fn build(self) -> Result<Offer, ()> {
+		match self.offer.amount {
+			Some(Amount::Bitcoin { amount_msats }) => {
+				if amount_msats > MAX_VALUE_MSAT {
+					return Err(());
+				}
+			},
+			Some(Amount::Currency { .. }) => unreachable!(),
```

Comment:
FWIW, it already fails in the next PR when we have actual errors to differentiate (https://github.com/lightningdevkit/rust-lightning/pull/1726/commits/cadc69c8b5cdb33548b36e38228693cfa62fe936). Do you mind if we just leave this as is until then to avoid more rebasing?